README.md
# `tir`
Have [**time.ir**](http://time.ir) in shell!


![tir-screenshot](https://user-images.githubusercontent.com/20663776/48029334-dbb70d80-e145-11e8-86ce-20746f1d89d2.png)


## Installation
#### Download
Using `git`:  
```sh
~ $ git clone --depth 1 --branch 19.5.15 https://github.com/Pouriya-Jahanbakhsh/tir && cd tir
~/tir $
```
Using `wget`:  
```sh
~ $ wget https://github.com/Pouriya-Jahanbakhsh/tir/archive/19.5.15.tar.gz && tar xf 19.5.15.tar.gz && cd tir-19.5.15
~/tir-19.5.15 $
```

#### Dependencies
First you need to have **Python 3.^5** installed. Most of other dependencies are by default installed on most Linuxes.  
* **setuptools**  
    On most linux distributions:  
    ```sh
    ~ $ sudo apt install python3-setuptools
    ```  
    On FreeBSD:  
    ```sh
    ~ # pkg install py36-setuptools
    ```
* **requests**  
    On most linux distributions:  
    ```sh
    ~ $ sudo apt install python3-requests
    ```  
    On FreeBSD:  
    ```sh
    ~ # pkg install py36-requests
    ```  
* **lxml**  
    On most linux distributions:  
    ```sh
    ~ $ sudo apt install python3-lxml
    ```  
    On FreeBSD:  
    ```sh
    ~ # pkg install py36-lxml
    ```  
    Note that `lxml` itself needs `libxml2` and `libxslt` to compile.

Note that on `FreeBSD` it's better to link your `python3.*` to `python3`:
```sh
~ # ln -s /usr/local/bin/python3.6 /usr/local/bin/python3
```
Or edit `Makefile` and first line of `./bin/crawler.py` and place your own Python 3 command

#### install `tir` itself
On Linux distributions run:
```sh
~/path/to/tir $ sudo make install
```  
On `FreeBSD` you need `gmake` instead of `make`:
```sh
~/path/to/tir # gmake install
```  

Then you have `tir` command everywhere:
```text
/x/y/z $ tir
Emruz: 3-Shanbeh  15   Aban(08)   1397   Pa'eez  
Today:  Tuesday   06 November(11) 2018   Autumn  
System time: 00:04:03
 ________  ________  ________  ________  ________  ________  ________
| Shanbe ||  Yek   ||   Do   ||   Se   || Chahar ||  Panj  || Jom'eh |
 ________  ________  ________  ________  ________  ________  ________ 
|        ||        ||        ||        ||        ||        ||        |
|   28   ||   29   ||   30   ||   01   ||   02   ||   03   ||   04   |
| 10  20 || 11  21 || 12  22 || 13  23 || 14  24 || 15  25 || 16  26 |
|________||________||________||________||________||________||________|
 ________  ________  ________  ________  ________  ________  ________ 
|        ||        ||        ||        ||        ||        ||        |
|   05   ||   06   ||   07   ||   08   ||   09   ||   10   ||   11   |
| 17  27 || 18  28 || 19  29 || 20  30 || 21  31 || 22  01 || 23  02 |
|________||________||________||________||________||________||________|
 ________  ________  ________  ________  ________  ________  ________ 
|        ||        ||        ||        ||        ||        ||        |
|   12   ||   13   ||   14   ||   15   ||   16   ||   17   ||   18   |
| 24  03 || 25  04 || 26  05 || 27  06 || 28  07 || 29  08 || 01  09 |
|________||________||________||________||________||________||________|
 ________  ________  ________  ________  ________  ________  ________ 
|        ||        ||        ||        ||        ||        ||        |
|   19   ||   20   ||   21   ||   22   ||   23   ||   24   ||   25   |
| 02  10 || 03  11 || 04  12 || 05  13 || 06  14 || 07  15 || 08  16 |
|________||________||________||________||________||________||________|
 ________  ________  ________  ________  ________  ________  ________ 
|        ||        ||        ||        ||        ||        ||        |
|   26   ||   27   ||   28   ||   29   ||   30   ||   01   ||   02   |
| 09  17 || 10  18 || 11  19 || 12  20 || 13  21 || 14  22 || 15  23 |
|________||________||________||________||________||________||________|

Powered by http://time.ir
```

## Uninstallation
On Linux distributions run:
```sh
~/path/to/tir $ sudo make uninstall
```  
On `FreeBSD` you need `gmake` instead of `make`:
```sh
~/path/to/tir # gmake uninstall
```  

## Arguments
```sh
root@codefather:~/tir $ tir -h
```
```text
Usage: tir [options]

Options:
  -h, --help       show this help message and exit
  -s, --solar      Does not show solar date
  -g, --gregorian  Does not show gregorian date
  -c, --calendar   Does not show calendar
  -t, --time       Does not show time
  -C, --color      Does not show colored text
  -q, --quote      Does not notify for quote
  -H, --holidays   Does not notify for holidays
  -a, --about      shows program's description and exits
```
So for example `tir -s -g -c -C -q -H` will result:
```sh
Emruz: 3-Shanbeh  15   Aban(08)   1397   Pa'eez  

Powered by http://time.ir
```

# Contributing
I love pull requests from everyone ! For more info see `CONTRIBUTING.md` file.


CHANGELOG.md
### 19.5.12
* **Fix(es):**

    * get rid of 'topWrapper' tag  

* **Refactor(s):**

    * month name format  

Generated at 2019-05-13 00:01


tir/utils.py
"""
tir.utils
~~~~~~~~~~~~~~
This module provides utility functions that are used within Tir
"""


def transform_date(data):
    """ """
    # for example it may be 'چهارشنبه - ۹ آبان ۱۳۹۷'
    data = data.split('-')
    weekday = data[0].strip()
    if not weekday.isalpha(): # it's farsi
        weekday = transform_weekday(weekday)
    date = transform_date_with_string_month(data[1].strip())
    return (weekday, date)

def transform_numerical_date(data):
    """ """
    # for example '1397/12/6' or 2018-2-25 which is my birthday :)
    if data.find('/') != -1:
        type_ = 'solar'
    else: # assume gregorian
        type_ = 'gregorian'
    data = data.strip().replace('/', '-').split('-')
    (year, month, day) = (transform_number(data[0])
                         ,transform_number(data[1])
                         ,transform_number(data[2]))
    season = find_season(int(month), type_)
    return (year, season, month, day)

def transform_date_with_string_month(date):
    """ """
    date = date.split()
    if date[2].isalpha():
        (year, day, month) = (date[0], date[1], date[2])
    else: # It's farsi
        (day, month, year) = (transform_number(date[0])
                             ,transform_month(date[1])
                             ,transform_number(date[2]))
    return (year, month, day)

def transform_weekday(weekday):
    """ """
    if len(weekday) < 4: # shorter weekdays are شنبه or جمعه
        raise ValueError('unknown weekday {!r}'.format(weekday))
    farsi_weekdays_table = [('Shanbeh',   1588)  # ش
                           ,('1-Shanbeh', 1740)  # ی
                           ,('2-Shanbeh', 1583)  # د
                           ,('3-Shanbeh', 1587)  # س
                           ,('4-Shanbeh', 1670)  # چ
                           ,('5-Shanbeh', 1662)  # پ
                           ,('Jom\'eh',   1580)] # ج
    weekday_char1_unicode_number = ord(weekday[0])
    for farsi_weekday, char1_unicode_number in farsi_weekdays_table:
        if weekday_char1_unicode_number == char1_unicode_number:
            return farsi_weekday
    raise ValueError('unknown weekday {!r}'.format(weekday))

def transform_month(month):
    """ """
    if len(month) < 3: # Solar months shoudl contain at least 3 characterss
        raise ValueError('unknown solar-hijri month {!r}'.format(month))
    farsi_month_table = [('Farvardin',   (1601, None))  # ف
                        ,('Ordibehesht', (1575, 1585))  # ار
                        ,('Khordad',     (1582, None))  # خ
                        ,('Tir',         (1578, None))  # ت
                        ,('Mordad',      (1605, 1585))  # مر
                        ,('Shahrivar',   (1588, None))  # ش
                        ,('Mehr',        (1605, 1607))  # مه
                        ,('Aban',        (1575, 1576))  # اب
                        ,('Azar',        (1575, 1584))  # اذ
                        ,('Dey',         (1583, None))  # د
                        ,('Bahman',      (1576, None))  # ب
                        ,('Esfand',      (1575, 1587))] # اس
    (char1_number, char2_number) = (ord(month[0]), ord(month[1]))
    if char1_number == 1570:
        char1_number = 1575 # transform آ to ا (A without hat :D)
    for name, char_numbers in farsi_month_table:
        if char_numbers[0] == char1_number:
            if (not char_numbers[1] or char_numbers[1] == char2_number):
                return name
    raise ValueError('unknown solar month {!r}'.format(month))

def transform_number(numbers):
    """ Converts Persian and Arabic numbers to English numbers"""
    new_format = ''

    mapping_dictionary = {
            # Persian numbers
            '۰': '0', '۱': '1', '۲': '2', '۳': '3', '۴': '4',
            '۵': '5', '۶': '6', '۷': '7', '۸': '8', '۹': '9',

            # Arabic numbers
            '.': '0', '١': '1', '٢': '2', '٣': '3', '٤': '4',
            '٥': '5', '٦': '6', '٧': '7', '٨': '8', '٩': '9',
            }

    for number in numbers:
        result = mapping_dictionary.get(number, None)
        if result is None:
            raise ValueError('unknown farsi number {!r}'.format(number))
        new_format += result
    return new_format.zfill(2)



tir/tir.py
import requests
import lxml.html
from collections import namedtuple


# named tuples:
# used for parsed data and themes

Date = namedtuple('Date', ['year'        # e.g. '2018'
                          ,'season'      # '01'-'03' which means 3rd season
                          ,'season_name' # e.g. 'Bahar' or 'Spring'
                          ,'month'       # '01'-'12'
                          ,'month_name'  # e.g. 'Shahrivar' or 'June'
                          ,'day'         # '01'-'31'
                          ,'weekday'])   # e.g. 'Shanbeh' or 'Saturday'

Day = namedtuple('Day', ['is_disabled' # boolean
                        ,'is_today'    # boolean
                        ,'is_holiday'  # boolean
                        ,'solar'       # '01'-'31'
                        ,'gregorian'   # '01'-'31'
                        ,'qamari'])    # '01'-'31'

Time = namedtuple('Time', ['hour'     # '00'-'23'
                          ,'minute'   # '00'-'59'
                          ,'second']) # '00'-'59'

Quote = namedtuple('Quote', ['author', 'text']) # string


# each item in theme should be 2-sized tuple
# first element is start and second element is stop
# for example defining ('\033[1;31m', '\033[0m') as value of 'disabled'
#  in CalendarTheme, means that print disabled days in Red color.
CalendarTheme = namedtuple('CalendarTheme', ['disabled'
                                            ,'holiday'
                                            ,'today'
                                            ,'normal'
                                            ,'solar'
                                            ,'other_days'])

DateTheme = namedtuple('Date', ['year'
                               ,'seasons'
                               ,'month'
                               ,'month_name'
                               ,'weekday'
                               ,'day'])

TimeTheme = namedtuple('TimeTheme', ['hour'
                                    ,'minute'
                                    ,'second'])

# parser functions:

def transform_date(data):
    # for example it may be 'چهارشنبه - ۹ آبان ۱۳۹۷'
    data = data.split('-')
    weekday = data[0].strip()
    if not is_a_to_z(weekday): # it's farsi
        weekday = transform_weekday(weekday)
    date = transform_date_with_string_month(data[1].strip())
    return (weekday, date)


def transform_numeral_date(data):
    # for example '1397/12/6' or 2018-2-25 which is my birthday :)
    if data.find('/') != -1:
        _type = 'solar'
    else: # assume gregorian
        _type = 'gregorian'
    data = data.strip().replace('/', '-').split('-')
    (year, month, day) = (transform_number(data[0])
                         ,transform_number(data[1])
                         ,transform_number(data[2]))
    season = find_season(int(month), _type)
    return (year, season, month, day)


def transform_date_with_string_month(date):
    date = date.split()
    if is_a_to_z(date[2]):
        (year, day, month) = (date[0], date[1], date[2])
    else: # It's farsi
        (day, month, year) = (transform_number(date[0])
                             ,transform_month(date[1])
                             ,transform_number(date[2]))
    return (year, month, day)


def transform_weekday(weekday):
    if len(weekday) < 4: # shorter weekdays are شنبه or جمعه
        raise ValueError('unknown weekday {!r}'.format(weekday))
    farsi_weekdays_table = [('Shanbeh',   1588)  # ش
                           ,('1-Shanbeh', 1740)  # ی
                           ,('2-Shanbeh', 1583)  # د
                           ,('3-Shanbeh', 1587)  # س
                           ,('4-Shanbeh', 1670)  # چ
                           ,('5-Shanbeh', 1662)  # پ
                           ,('Jom\'eh',   1580)] # ج
    weekday_char1_unicode_number = ord(weekday[0])
    for farsi_weekday, char1_unicode_number in farsi_weekdays_table:
        if weekday_char1_unicode_number == char1_unicode_number:
            return farsi_weekday
    raise ValueError('unknown weekday {!r}'.format(weekday))


def find_season(month_number, _type):
    season_table = [('Bahar',    'Spring')
                   ,('Tabestan', 'Summer')
                   ,('Pa\'eez',  'Autumn')
                   ,('Zemestan', 'Winter')]
    if _type == 'solar':
        season_name_offset = 0
        if 0 < month_number < 4:
            season_number = '01'
        elif 3 < month_number < 7:
            season_number = '02'
        elif 6 < month_number < 10:
            season_number = '03'
        else: # 10, 11, 12
            season_number = '04'
    else: # assume gregorian
        season_name_offset = 1
        if 2 < month_number < 6:
            season_number = '01'
        elif 5 < month_number < 9:
            season_number = '02'
        elif 8 < month_number < 12:
            season_number = '03'
        else: # 12, 01, 02
            season_number = '04'
    season_name = season_table[int(season_number)-1][season_name_offset]
    return (season_name, season_number)


def transform_number(number):
    number2 = ''
    for char in number:
        unicode_number = ord(char)
        if 1775 < unicode_number < 1786: # ۰-۹ farsi
            number2 += chr(unicode_number - 1728) # transform farsi ۰-۹ to 0-9
            continue
        if 1631 < unicode_number < 1642: # ۰-۹ arabic
            number2 += chr(unicode_number - 1584) # transform arabic ۰-۹ to 0-9
            continue
        if 47 < ord(char) < 58: # 0-9
            number2 += char
            continue
        raise ValueError('unknown farsi number {!r}'.format(number))
    if len(number2) == 1:
        number2 = '0' + number2
    return number2


def transform_month(month):
    if len(month) < 2: # Solar months shoudl contain at least 2 characterss
        raise ValueError('unknown solar-hijri month {!r}'.format(month))
    farsi_month_table = [('Farvardin',   (1601, None))  # ف
                        ,('Ordibehesht', (1575, 1585))  # ار
                        ,('Khordad',     (1582, None))  # خ
                        ,('Tir',         (1578, None))  # ت
                        ,('Mordad',      (1605, 1585))  # مر
                        ,('Shahrivar',   (1588, None))  # ش
                        ,('Mehr',        (1605, 1607))  # مه
                        ,('Aban',        (1575, 1576))  # اب
                        ,('Azar',        (1575, 1584))  # اذ
                        ,('Dey',         (1583, None))  # د
                        ,('Bahman',      (1576, None))  # ب
                        ,('Esfand',      (1575, 1587))] # اس
    (char1_number, char2_number) = (ord(month[0]), ord(month[1]))
    if char1_number == 1570:
        char1_number = 1575 # transform آ to ا (A without hat :D)
    for name, char_numbers in farsi_month_table:
        if char_numbers[0] == char1_number:
            if (not char_numbers[1] or char_numbers[1] == char2_number):
                return name
    raise ValueError('unknown solar month {!r}'.format(month))


def is_a_to_z(data):
    for char in data:
        number = ord(char)
        if 96 < number < 123: # a-z
            continue
        if 64 < number < 91:  # A-Z
            continue
        return False
    return True

# search funcion and its utilities which used for parsed HTML page:

def search(element, tag, attr, val):
    def _has_attr(sub_element):
        for attr2, val2 in sub_element.attrib.items():
            if type(attr) == tuple and type(val) == tuple:
                if attr2.find(attr[0]) != -1 and val2.find(val[0]) != -1:
                    return True
            elif type(attr) == tuple:
                if attr2.find(attr[0]) != -1 and val2 == val:
                    return True
            elif type(val) == tuple:
                if attr2 == attr and val2.find(val[0]) != -1:
                    return True
            else:
                if attr2 == attr and val2 == val:
                    return True
        return False

    def _search(element, tag, attr, val):
        for sub_element in element.getchildren():
            if sub_element.tag == tag:
                if attr == None:
                    return sub_element
                if _has_attr(sub_element):
                    return sub_element
            result = _search(sub_element, tag, attr, val)
            if result == None:
                continue
            return result
        return None

    element = _search(element, tag, attr, val)
    if element == None:
        raise _TagNotFound(tag, attr, val)
    return element

class _TagNotFound(Exception):

    def __init__(self, tag, attr=None, value=None):
        text = 'could not found HTML tag {!r} '.format(tag)
        if attr:
            text += 'with attribute '
            if type(attr) == tuple:
                text += 'which should contain {!r} '.format(attr[0])
            else:
                text += '{!r} '.format(attr)
            text += 'and value '
            if type(value) == tuple:
                text += 'which should contain {!r}'.format(value[0])
            else:
                text += '{!r}'.format(value)
        Exception.__init__(self, text)

# A class which makes request for fetching HTML page:

class Request:

    def __init__(self
                ,url='http://time.ir'
                ,user_agent = 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:50.0) Gecko/20100101 Firefox/50.0'
                ,headers = {'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'}):
        self.url = url
        for (key, _) in headers.items():
            if key.lower() == 'user-agent':
                break
        else:
            headers['user-agent'] = user_agent
        self.headers = headers


    def get(self):
        request = requests.get(self.url, headers=self.headers)
        body = request.text
        assert(len(body) > 10240) # It's normals size is about 80K, but wee need at least 10K to process
        return body

# An HTML parser which accepts some transformers, and after parsing HTML
#  page, runs each transformer with parsed data


class HTMLParser:
    
    def __init__(self, text, transformers):
        self.html = lxml.html.fromstring(text)
        self.transformers = transformers

    def parse(self):
        transform_data = {}
        for name, transformer in self.transformers.items():
            transform_data[name] = transformer(self.html)
        return transform_data
